python/triton/compiler.py

```diff
@@ -1546,43 +1561,57 @@ def format_of(ty):
   return ptr_info;
 }}
 
-static PyObject* launch(PyObject* self, PyObject* args) {{
-  int gridX, gridY, gridZ;
-  uint64_t _stream;
-  uint64_t _function;
-  int num_warps;
-  int shared_memory;
-  PyObject *launch_enter_hook = NULL;
-  PyObject *launch_exit_hook = NULL;
-  PyObject *compiled_kernel = NULL;
-  {' '.join([f"{_extracted_type(ty)} _arg{i}; " for i, ty in signature.items()])}
-  if(!PyArg_ParseTuple(args, \"{format}\", &gridX, &gridY, &gridZ, &num_warps, &shared_memory, &_stream, &_function, &launch_enter_hook, &launch_exit_hook, &compiled_kernel, {', '.join(f"&_arg{i}" for i, ty in signature.items())})) {{
-    return NULL;
+static PyObject* launch(PyObject* self, PyObject* const* args, Py_ssize_t nargs, PyObject* kwnames) {{
+  int gridX = (int) PyLong_AsLong(args[0]);
+  int gridY = (int) PyLong_AsLong(args[1]);
+  int gridZ = (int) PyLong_AsLong(args[2]);
+  int num_warps = (int) PyLong_AsLong(args[3]);
+  int shared_memory = (int) PyLong_AsLong(args[4]);
+
+  uint64_t _stream = PyLong_AsUnsignedLongLong(args[5]);
+  uint64_t _function = PyLong_AsUnsignedLongLong(args[6]);
+
+  PyObject *launch_enter_hook = args[7];
+  PyObject *launch_exit_hook = args[8];
+  PyObject *compiled_kernel = args[9];
+
+  {" ".join([f"PyObject* _arg{i} = args[{10+i}];" if ty[0] == "*" else "" for i, ty in signature.items()])};
+  {" ".join([f"{ty_to_cpp(ty)} _arg{i} = ({ty_to_cpp(ty)}) {ty_to_cast_method(ty)}(args[{10+i}]);" if ty[0] != "*" else "" for i, ty in signature.items()])};
+
+  PyObject *py_args = NULL;
+  if(launch_enter_hook != Py_None || launch_exit_hook != Py_None) {{
+    py_args = PyTuple_New(nargs);
+    for(Py_ssize_t i = 0; i < nargs; ++i) {{
+        Py_INCREF(args[i]);
```

Comment:
does this need to get decremented at some point?

Reply:
I think once we decrement the reference count of py_args, and its reference count reaches 0, its elements would get decremented automatically.

Reply:
The set-item "seals" the reference here:
https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SetItem

Reply:
I don't understand this change? So we cannot pass `args` to the hook?

Reply:
Ah, I get it...`args` has been changed to a pointer of another pointer.

Reply:
`args` is `PyObject* const*` whereas the `PyObject_CallObject` takes `PyObject*`

Reply:
Could `*args` work?